lib/hd/keyorigin.js

```diff
@@ -0,0 +1,174 @@
+'use strict';
+
+const assert = require('assert');
+const common = require('./common');
+const bio = require('bufio');
+const hash160 = require('bcrypto/lib/hash160');
+const WalletKey = require('../wallet/walletkey');
+
+/**
+ * Helper class to represent hd key path for arbitrary wallets.
+ * @property {Number} fingerPrint - master key fingerprint (uint32)
+ * @property {Array} path - bip32 derivation path in uint32 array
+ */
+class KeyOriginInfo {
+    constructor(options) {
+        this.fingerPrint = -1;
+        this.path = [];
+        if (options) {
+            this.fromOptions(options);
+        }
+    }
+
+    fromOptions(options) {
+        assert(options, 'requires options');
+        if (options.fingerPrint) {
+            assert(
+                (options.fingerPrint >>> 0) === options.fingerPrint,
+                'fingerPrint must be uint32'
+            );
+            this.fingerPrint = options.fingerPrint;
+        }
+        if (options.path) {
+            if (Array.isArray(options.path)) {
+                assert(
+                    options.path.every(p => (p >>> 0) === p),
+                    'all path index must be uint32'
+                );
+                this.path = options.path;
+            } else {
+                this.path = common.parsePath(options.path, true);
+            }
+        }
+        return this;
+    }
+
+    static fromOptions(options) {
+        return new this().fromOptions(options);
+    }
+
+    equals(keyInfo) {
+        assert(KeyOriginInfo.isKeyOriginInfo(keyInfo));
+        if (this.fingerPrint !== keyInfo.fingerPrint)
+            return false;
+        for (const i in this.path) {
+            if (this.path[i] !== keyInfo.path[i])
+                return false;
+        }
+        return true;
+    }
+
+    inspect() {
+        return this.format();
+    }
+
+    format() {
+        let path = 'm';
+        for (const p of this.path) {
+            const hardened = (p & common.HARDENED) ? '\'' : '';
+            path += `/${p & 0x7fffffff}${hardened}`;
+        }
+        return {
+            fingerPrint: this.fingerPrint,
+            path
+        };
+    }
+
+    static fromRaw(data) {
+        return new this().fromRaw(data);
+    }
+
+    fromRaw(data) {
+        return this.fromReader(bio.read(data));
+    };
+
+    static fromReader(br) {
+        return new this().fromReader(br);
+    }
+
+    fromReader(br) {
+        this.fingerPrint = br.readU32BE();
+        while (br.left()) {
+            this.path.push(br.readU32());
+        }
+        return this;
+    }
+
+    toRaw() {
+        const size = this.getSize();
+        return this.toWriter(bio.write(size)).render();
+    }
+
+    toWriter(bw) {
+        bw.writeU32BE(this.fingerPrint);
+        for (const p of this.path) {
+            bw.writeU32(p);
+        }
+        return bw;
+    }
+
+    toJSON() {
+        return {
+            fingerPrint: this.fingerPrint,
+            path: this.path
```

Comment:
Don't have a strong opinion but we should return:

1. fingerprint as a hex string
2. path as a string like 'm/48'/...'

Reply:
Resolved